### ISSUE TITLE
Crash with Quilt due to config directory override

### ISSUE THREAD

**Issue by e-splat** [enjarai/shared-resources#6]
hello, i've updated sr to the latest version (1.2.2 to 1.2.3) and it crashes with FAPI (Fabric API) and QSL (Quilt Standard Libraries), i think is due to changes on the apis, please fix this

**Comment by enjarai** [comment]
Please send a log containing the crash, I'll see what I can do.

**Comment by e-splat** [comment]
https://0x0.st/o0jN.txt

**Comment by enjarai** [comment]
Ah, this is due to the config directory override which hooks directly into Fabric Loader. When Fabric isnt present, for example when using Quilt, it causes a crash.

Making it work for Quilt might be difficult, since this is primarily a Fabric mod, but I'll see if I can at least prevent it from crashing. For now Quilt isn't supported.

**Comment by AlexIIL** [comment]
Hi, quilt developer here. I think you can change quilt's config directory in the same way as fabric, except by modifying the `configDir` field inside of `org.quiltmc.loader.impl.QuiltLoaderImpl` instead.

However that's not the cause of the game crashing, which is actually a mixin apply error:

`
Caused by: org.spongepowered.asm.mixin.injection.throwables.InjectionError: Critical injection failure: Callback method sharedresources$getKeys(Lnet/minecraft/class_2487;)Lnet/minecraft/class_2487; in #shared-resources-mc18:shared-resources18.mixins.json:options.remembermytxt.GameOptionsMixin from mod shared-resources-mc18 failed injection check, (0/1) succeeded. Scanned 1 target(s). Using refmap shared-resources-mc18-refmap.json
`
Is this unrelated?

**Comment by enjarai** [comment]
Oh I see, I saw a `ClassNotFoundException` and completely neglected to scroll past that. Not my best troubleshooting moment. 😅 

I'm not having much success replicating this mixin error in my dev env though, using FAPI `0.58.0+1.18.2`. Do you have more insight into what could be going wrong here?

Quilt support is not a high priority for me unfortunately, but I wouldn't mind adding it. The problem is that Quilt classes arent accessible to me at compile time for obvious reasons, do you know any way around this when using reflection?

**Comment by AlexIIL** [comment]
> I'm not having much success replicating this mixin error in my dev env though, using FAPI 0.58.0+1.18.2. Do you have more insight into what could be going wrong here?

Ah, maybe it only happens in production? I can reproduce this with just fabric-loader `0.14.10`, fabric-api `0.67.1+1.19.2` and shared-resources `1.2.3`

As for the reflection - I think you can do it the long way round:
```java
if (FabricLoader.getInstance().isModLoaded("quilt_loader") {
  Class<?> qloader = Class.forName("org.quiltmc.loader.impl.QuiltLoaderImpl");
  Field configField = qloader.getDeclaredField("configDir");
  configField.setAccessible(true);
  configField.set(qloader.getDeclaredField("INSTANCE").get(null), new_config_directory);
} else {
  // Fabric-loader version
}
```

**Comment by enjarai** [comment]
Alright, I've found the problem (MixinExtras error messages are terrible) and it seems to be solved in 1.2.4, which was just published. It also includes Quilt compat similar to what you describe here. Thanks for the help with debugging. 😄
